server/integration/infinispan/src/main/java/org/jboss/as/clustering/infinispan/subsystem/CounterResource.java

```diff
@@ -0,0 +1,246 @@
+ *
+ * @author Vladimir Blagojevic
+ * @since 9.2
+ */
+public class CounterResource extends SimpleResourceDefinition {
+
+    //attributes
+    static final SimpleAttributeDefinition COUNTER_NAME = new SimpleAttributeDefinitionBuilder(ModelKeys.NAME,
+          ModelType.STRING, false)
+          .setXmlName(Attribute.NAME.getLocalName())
+          .setAllowExpression(false)
+          .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+          .build();
+
+    static final SimpleAttributeDefinition STORAGE = new SimpleAttributeDefinitionBuilder(ModelKeys.STORAGE,
+          ModelType.STRING, false)
+          .setXmlName(Attribute.STORAGE.getLocalName())
+          .setAllowExpression(false)
+          .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+          .setAllowedValues(Storage.VOLATILE.toString(), Storage.PERSISTENT.toString())
+          .setDefaultValue(new ModelNode().set(Storage.VOLATILE.toString()))
+          .build();
+
+    static final SimpleAttributeDefinition INITIAL_VALUE = new SimpleAttributeDefinitionBuilder(ModelKeys.INITIAL_VALUE,
+          ModelType.LONG, true)
+          .setXmlName(Attribute.INITIAL_VALUE.getLocalName())
+          .setAllowExpression(false)
+          .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+          .setDefaultValue(new ModelNode().set(0))
+          .build();
+
+    static final SimpleAttributeDefinition VALUE = new SimpleAttributeDefinitionBuilder(ModelKeys.VALUE,
+          ModelType.LONG, true)
+          .setXmlName(Attribute.VALUE.getLocalName())
+          .setAllowExpression(false)
+          .setFlags(AttributeAccess.Flag.RESTART_RESOURCE_SERVICES)
+          .build();
+
+    static final AttributeDefinition[] ATTRIBUTES = { COUNTER_NAME, STORAGE, INITIAL_VALUE };
+
+    // operations
+
+    private static final OperationDefinition COUNTER_RESET = buildOperation("counter-reset").build();
+    private static final OperationDefinition COUNTER_INCREASE = buildOperation("counter-increase").build();
+    private static final OperationDefinition COUNTER_DECREASE = buildOperation("counter-decrease").build();
+
+    private final boolean runtimeRegistration;
+
+    public CounterResource(PathElement pathElement, ResourceDescriptionResolver descriptionResolver,
+            AbstractAddStepHandler addHandler, OperationStepHandler removeHandler, boolean runtimeRegistration) {
+        super(pathElement, descriptionResolver, addHandler, removeHandler);
+        this.runtimeRegistration = runtimeRegistration;
+    }
+
+    @Override
+    public void registerAttributes(ManagementResourceRegistration resourceRegistration) {
+        super.registerAttributes(resourceRegistration);
+        final OperationStepHandler writeHandler = new ReloadRequiredWriteAttributeHandler(ATTRIBUTES);
+        for (AttributeDefinition attr : ATTRIBUTES) {
+            resourceRegistration.registerReadWriteAttribute(attr, null, writeHandler);
```

Comment:
can you explain what is happening here? the counter-name, storage and initial-value can't be changed and they are set when you define a counter.

Reply:
Here you have specified the `readHandler` as null, is that correct?

My understanding from what @pruivo says is that the values are read only? If so, we can use `registerReadOnlyAttribute`

Reply:
I tried it @pruivo and for some reason I get 
```
[domain@localhost:9990 counters=COUNTERS] cd strong-counter=c4
[domain@localhost:9990 strong-counter=c4] ls
WFLYCTL0063: Composite operation was rolled back
[domain@localhost:9990 strong-counter=c4] 
```

Reply:
@vblagoje What was the output on the server?

Reply:
@ryanemerson , nothing, that's was weird

Reply:
this is a no-go for me. the model is updated with the new values but the configuratio stays the same.

Reply:
@pruivo what exactly are you doing to update the model? I am doing this:

```
[domain@localhost:9990 counters=COUNTERS] pwd
/profile=clustered/subsystem=datagrid-infinispan/cache-container=clustered/counters=COUNTERS
[domain@localhost:9990 counters=COUNTERS] ./strong-counter=c89:add(
initial-value=  lower-bound=  name=  storage=  upper-bound=  
[domain@localhost:9990 counters=COUNTERS] ./strong-counter=c89:add(name=c89, initial-value=0, lower-bound=0, storage=VOLATILE,upper-bound=20)
{
    "outcome" => "success",
    "result" => undefined,
    "server-groups" => {"cluster" => {"host" => {"master" => {
        "server-one" => {"response" => {"outcome" => "success"}},
        "server-two" => {"response" => {"outcome" => "success"}}
    }}}}
}
[domain@localhost:9990 counters=COUNTERS] cd strong-counter=c
c1  c89  
[domain@localhost:9990 counters=COUNTERS] cd strong-counter=c89
/  :  
[domain@localhost:9990 counters=COUNTERS] cd strong-counter=c89
[domain@localhost:9990 strong-counter=c89] :write-attribute(name=upper-bound, value=30)
{
    "outcome" => "success",
    "result" => undefined,
    "server-groups" => {"cluster" => {"host" => {"master" => {
        "server-one" => {"response" => {
            "outcome" => "success",
            "response-headers" => {
                "operation-requires-reload" => true,
                "process-state" => "reload-required"
            }
        }},
        "server-two" => {"response" => {
            "outcome" => "success",
            "response-headers" => {
                "operation-requires-reload" => true,
                "process-state" => "reload-required"
            }
        }}
    }}}}
}
[domain@localhost:9990 strong-counter=c89] ls
initial-value=0   lower-bound=0     name=c89          storage=VOLATILE  upper-bound=30    

[domain@localhost:9990 strong-counter=c89]
```
It all works as expected, once server is rebooted, the new value is loaded into configuration.

Reply:
the volatile counter is destroyed in the reboot. so you are creating a new counter. a persistent counter isn't and the configuration isn't updated...

Reply:
and there is a small issue. it destroys *all* the volatile counters!

Reply:
But how can I circumvent this with current Counter API, not sure what I can do to handle this properly? If I destroy counter (with remove) then I can not save its value. If I don't destroy it (remove it) then I can not update its configuration on reboot...it is chicken-egg issue, right? Suggestions?

Reply:
that's why they should be read-only for now. when the `undefine()` is implemented, we can support to write the attributed!

Reply:
Ok @pruivo will investigate that impl now. Thanks